File: ExamenPy/ZooVisit.py
import random
from Employee import Employee
from Visitant import Visitant

class ZooVisit:
    __visit_date = ""
    __visitant_list = []
    __amount = 0.0
    __num_child = 0
    __num_adult = 0
    __ID = ""
    
    def __init__(self, guide, visit_date, visitant_list):
        self.__guide = guide
        self.__guide.set_active_employee(True)
        self.__visit_date = visit_date
        self.__visitant_list = visitant_list
        for visitant in self.__visitant_list:
            visitant.set_active_visitant(True)
        self.calculate_num_adult()
        self.calculate_num_child()
        self.calculate_amount()
        self.__ID = self.generate_id()

    # GETTERS AND SETTERS
    def get_guide(self):
        return self.__guide
    def get_visitant_list(self):
        return self.__visitant_list
    def get_amount(self):
        return self.__amount
    def get_visit_date(self):
        return self.__visit_date
    def get_num_child(self):
        return self.__num_child
    def get_num_adult(self):
        return self.__num_adult
    def get_id(self):
        return self.__ID
    def set_id(self, ID):
        self.__ID = ID 

    def activate_visitants(self):
        for visitant in self.__visitant_list:
            visitant.set_active_visitant(True)

    def calculate_num_adult(self):
        for visitant in self.__visitant_list:
            if visitant.is_adult():
                self.__num_adult += 1

    def calculate_num_child(self):
        self.__num_child = len(self.__visitant_list) - self.__num_adult

    def calculate_amount(self):
        for visitant in self.__visitant_list:
            visitant.add_visit()
            if visitant.is_adult():
                if visitant.get_num_visits() % 5 == 0:
                    self.__amount += 80
                else:
                    self.__amount += 100
            else:
                self.__amount += 50

    def generate_id(self):
        cad = "ZV"
        letters_bank = "ABCDEFGHIJKLMNÑOPQRSTUVWXYZ"
        cad += random.choice(letters_bank)
        i = 0
        while i < 4:                
            cad += str(random.randint(0,9))
            i += 1
        return cad

File: ExamenJava/ZooVisit.java
import java.util.ArrayList;

public class ZooVisit {
    private Employee guide;
    private ArrayList<Visitant> visitantList;
    private String visitDate;
    private int numChild, numAdult;
    private double amount;
    private String ID;

    public ZooVisit(Employee guide, String visitDate, ArrayList<Visitant> visitantList, String ID) {
        this.guide = guide;
        guide.setActiveEmployee(true);
        this.visitDate = visitDate;
        this.visitantList = visitantList;
            for (Visitant visitant : visitantList) {
                visitant.setActiveVisitant(true);
            } 
        calculateNumAdult();
        calculateNumChild();
        calculateAmount();
        this.ID = ID;
    }

    //GETTERS AND SETTERS
    public Employee getGuide() {
        return guide;
    }
    public ArrayList<Visitant> getVisitantList() {
        return visitantList;
    }
    public double getAmount() {
        return amount;
    }
    public String getVisitDate() {
        return visitDate;
    }
    public int getNumChild() {
        return numChild;
    }
    public int getNumAdult() {
        return numAdult;
    }
    public String getID(){
        return this.ID;
    }

    public void activateVisitants(){
        for (Visitant visitant : this.visitantList) {
            visitant.setActiveVisitant(true);
        }
    }

    private void calculateNumAdult() {
        for (Visitant visitant : this.visitantList) {
            if (visitant.isAdult()) {
                this.numAdult++;
            }
        }
    }

    private void calculateNumChild() {
        this.numChild = this.visitantList.size() - numAdult;
    }

    private void calculateAmount() {
        for (Visitant visitant : this.visitantList) {
            visitant.addVisit();
                if(visitant.isAdult()){
                    if(visitant.getNumVisits() % 5 == 0){
                        this.amount += 80;
                    }
                    else{
                        this.amount += 100;
                    }
                }
                else{
                    this.amount += 50;
                }
        }
    }
}